src/lib/address_resolve/AddressResolve_DefaultImpl.cpp

```diff
@@ -112,6 +100,65 @@ NodeLookupAction NodeLookupHandle::NextAction(System::Clock::Timestamp now)
     return NodeLookupAction::KeepSearching();
 }
 
+bool NodeLookupResults::UpdateResults(const ResolveResult & result)
+{
+    auto newScore = Dnssd::IPAddressSorter::ScoreIpAddress(result.address.GetIPAddress(), result.address.GetInterface());
+
+    uint8_t insertAtIndex = 0;
+    for (; insertAtIndex < kNodeLookupResultsLen; insertAtIndex++)
+    {
+        if (insertAtIndex >= count)
+        {
+            // This is a new entry.
+            break;
+        }
+
+        auto & oldAddress = results[insertAtIndex].address;
+        auto oldScore     = Dnssd::IPAddressSorter::ScoreIpAddress(oldAddress.GetIPAddress(), oldAddress.GetInterface());
+        if (newScore > oldScore)
+        {
+            // This is a score update, it will replace a previous entry.
+            break;
+        }
+    }
+
+    if (insertAtIndex == kNodeLookupResultsLen)
+    {
+        return false;
+    }
+
+    // Move the following valid entries one level down.
+    for (auto i = count; i > insertAtIndex; i--)
+    {
+        if (i >= kNodeLookupResultsLen)
+        {
+            continue;
+        }
+
+        results[i] = results[i - 1];
+    }
```

Comment:
Please add unit tests to guarantee this code works. Whenever you have code that moves entries in place, there is potential for indexing mistakes. I looked hard at the indexing, and I can't convince myself there isn't a corner case or off-by-one in the method